File: src/main/java/com/example/oauth/util/CookieUtils.java
package com.example.oauth.util;

import lombok.experimental.UtilityClass;

import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.Optional;

@UtilityClass
public class CookieUtils {

    public Optional<Cookie> getCookie(HttpServletRequest request, String key) {
        Cookie[] cookies = request.getCookies();
        if (cookies != null) {
            for (Cookie cookie : cookies) {
                if (cookie.getName().equals(key)) {
                    return Optional.of(cookie);
                }
            }
        }
        return Optional.empty();
    }

    public void setCookie(HttpServletResponse response, String key, String value, int maxAge) {
        Cookie cookie = new Cookie(key, value);
        cookie.setPath("/");
        cookie.setHttpOnly(true); // Frontend can not receive token by cookie
        cookie.setMaxAge(maxAge);
        response.addCookie(cookie);
    }

    public void deleteCookie(HttpServletRequest request, HttpServletResponse response, String name) {
        Cookie[] cookies = request.getCookies();
        if (cookies != null) {
            for (Cookie cookie : cookies) {
                if (cookie.getName().equals(name)) {
                    cookie.setValue("");
                    cookie.setPath("/");
                    cookie.setMaxAge(0);
                    response.addCookie(cookie);
                }
            }
        }
    }
}

File: src/main/java/com/example/oauth/util/CookieUtils.kt
package com.example.oauth.util

import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse
import java.lang.UnsupportedOperationException
import java.util.*
import javax.servlet.http.Cookie

class CookieUtils private constructor() {

    init {
        throw UnsupportedOperationException("This is a utility class and cannot be instantiated")
    }

    companion object {
        fun getCookie(request: HttpServletRequest, key: String): Cookie? {
            val cookies = request.cookies
            return cookies?.firstOrNull { it.name == key }
        }

        fun setCookie(response: HttpServletResponse, key: String, value: String, maxAge: Int) {
            val cookie = Cookie(key, value).apply {
                this.path = "/"
                this.isHttpOnly = true
                this.maxAge = maxAge
            }
            response.addCookie(cookie)
        }

        fun deleteCookie(request: HttpServletRequest, response: HttpServletResponse, name: String) {
            val cookies = request.cookies
            cookies?.firstOrNull { it.name == name }?.let {
                it.value = ""
                it.path = "/"
                it.maxAge = 0
                response.addCookie(it)
            }
        }
    }

}